src/gui/PianoView.cpp

```diff
@@ -106,10 +112,18 @@ PianoView::PianoView( QWidget * _parent ) :
 	{
 		s_blackKeyPressedPm = new QPixmap( embed::getIconPixmap( "black_key_pressed" ) );
 	}
+	if (s_whiteKeyDisabledPm == NULL)
+	{
+		s_whiteKeyDisabledPm = new QPixmap(embed::getIconPixmap("white_key_disabled"));
+	}
+	if (s_blackKeyDisabledPm == NULL)
+	{
+		s_blackKeyDisabledPm = new QPixmap(embed::getIconPixmap("black_key_disabled"));
+	}
 
-	setAttribute( Qt::WA_OpaquePaintEvent, true );
-	setFocusPolicy( Qt::StrongFocus );
-	setMaximumWidth( WhiteKeysPerOctave * NumOctaves * PW_WHITE_KEY_WIDTH );
+	setAttribute(Qt::WA_OpaquePaintEvent, true);
+	setFocusPolicy(Qt::StrongFocus);
+	setMaximumWidth(WhiteKeysPerOctave * NumKeys * PW_WHITE_KEY_WIDTH / KeysPerOctave);
```

Comment:
I think the white keys actually _are_ spaced evenly; the problem was probably re-introduced with the merge of recent master, since I see that the instrument window now has the correct size (it had a weird margin on the right side before). I just checked 1.2.1 and it splits the last white key on the right in half as well, so this would be better addressed in a separate PR, possibly as part of a larger PianoView refactor.

Reply:
All keys are spaced evenly (`m_keyLineHeight`), but are drawn differently.

Reply:
@Veratil I think you wandered into the wrong fairy tale, this is PianoView, not PianoRoll.. :D

The change in window size also caused the last key on the right to never be drawn fully, and the "last key" marker to be invisible in its default location, so I adjusted the scroll area to add space for one more key.

Reply:
D'oh! Sorry! 😁

Reply:
Oh, now looking back I finally understand what did Dom mean. I looked only at the instrument window, where the Piano View was always clipped to the right size, so it seemed correct. Resizing the instrument window so that the entire PianoView is visible at once shows the issue.

Reply:
I tried to find a simple way to count the white keys, but I always somehow ended up with a for loop for the last few keys (or adding another custom map like `WhiteKeys`, but for the opposite direction). The number of white keys is very unlikely to change (even MIDI 2.0, which we do not support, still uses only 127 keys), so it seems a waste to compute it every time or complicate the code, so I simply added the count as a constant to the `Piano` class.

I also moved the `WhiteKeysPerOctave` and `BlackKeysPerOctave` constants from `Note.h` to `Piano.h`, since they were only used by the Piano View, and key color is not really associated with a note, but with the instrument (piano), so the `Piano` class seems a more suitable place for it.